### PR TITLE
Www 214 dynamic filters

### PR DESCRIPTION
**Pull request by holingpoon**
This PR should fix the problem of Filter by Tags not changing when a user switches `audience` within a `season` for `staff-picks`. Please review. Thanks!

### CODE REVIEW

File: src/app/utils/utils.js [hunk 1]
```diff
@@ -66,6 +66,20 @@ function Utils() {
     return selectedTags;
   };
 
+  /**
+   * getFilters(filters)
+   * Generates a Map of (pick tag id => pick tag label) key-value pairs,
+   * returns an Object of this map.
+   * @param filters
```
Comment (EdwinGuzman) [inline src/app/utils/utils.js]: Update this to follow https://esdoc.org/ convention.

File: src/app/utils/utils.js [hunk 2]
```diff
@@ -66,6 +66,20 @@ function Utils() {
     return selectedTags;
   };
 
+  /**
+   * getFilters(filters)
```
Comment (EdwinGuzman) [inline src/app/utils/utils.js]: You're not really getting filters, you're mapping the filter array to id/label pairs, so maybe a better name?

File: src/app/components/BookFilters/BookFilters.jsx [hunk 3]
```diff
@@ -37,7 +35,8 @@ class BookFilters extends React.Component {
   }
 
   componentWillReceiveProps(nextProps) {
-    this.setState({ selectedFilters: nextProps.selectedFilters });
+    const newFilters = utils.getFilters(nextProps.filters);
```
Comment (EdwinGuzman) [inline src/app/components/BookFilters/BookFilters.jsx]: You can rename this variable to be `filters`. For creating objects, you can use the shortcut:
```
{
  selectedFilters: nextProps.selectedFilters,
  filters,
}
```

In the example above, the key is `filters` but since there's a variable with the same name, you do not have to do `filters: filters`.

File: src/app/components/BookFilters/BookFilters.jsx [hunk 4]
```diff
@@ -166,6 +167,7 @@ BookFilters.propTypes = {
   filters: PropTypes.array,
   selectableFilters: PropTypes.array,
   setSelectedFilter: PropTypes.func,
+  setSelectableFilters: PropTypes.func,
```
Comment (EdwinGuzman) [inline src/app/components/BookFilters/BookFilters.jsx]: Why did you add this here?

File: src/app/components/BookFilters/BookFilters.jsx [hunk 5]
```diff
@@ -17,15 +17,13 @@ class BookFilters extends React.Component {
...
-        id: filter.toLowerCase().split(' ').join('-'),
-        label: filter,
-      })),
+      filters: utils.getFilters(this.props.filters),
       selectedFilters: this.props.selectedFilters,
       focusId: '',
       disabled: false,
       showFilters: false,
       timeout: undefined,
+      selectableFilters: this.props.selectableFilters,
```
Comment (EdwinGuzman) [inline src/app/components/BookFilters/BookFilters.jsx]: Why did you add this here?

File: src/app/components/Application/Main.jsx [hunk 6]
```diff
@@ -182,15 +182,15 @@ class Main extends React.Component {
...
+
   render() {
     const { type } = this.props.picksData;
     const picksCount = this.getCount();
-
     return (
       <div className="nypl-row">
         <Sidebar
           filters={this.props.filters}
-          selectableFilters={this.state.selectableFilters}
```
Comment (EdwinGuzman) [inline src/app/components/Application/Main.jsx]: Since you're not using this anymore, you can remove this from the component.

File: src/app/components/BookFilters/BookFilters.jsx (hunk 3 above)
Comment (ktp242) [reply src/app/components/BookFilters/BookFilters.jsx]: or you can just do `filters: utils.getFilters(nextProps.filters)` if you don't need to assign `utils.getFilters(nextProps.filters)` anywhere else.

File: src/app/components/Application/Main.jsx (hunk 6 above)
Comment (holingpoon) [reply src/app/components/Application/Main.jsx]: Removed `this.state.selectableFilters` from `Main`'s constructor.

Comment (EdwinGuzman) [reply src/app/components/Application/Main.jsx]: Right, it's not just there... Check the file again to see other instances of it in the state that are not being used.

**Comment by holingpoon** [comment]
I have merged `development` into this PR and resolved the merge conflicts.

File: src/app/components/BookFilters/BookFilters.jsx [hunk 7]
```diff
@@ -37,7 +34,10 @@ class BookFilters extends React.Component {
   }
 
   componentWillReceiveProps(nextProps) {
-    this.setState({ selectedFilters: nextProps.selectedFilters });
+    this.setState({
+      selectedFilters: nextProps.selectedFilters,
+      filters: utils.getFiltersMapping(nextProps.filters)
```
Comment (ktp242) [inline src/app/components/BookFilters/BookFilters.jsx]: Missing a trailing comma here

File: test/unit/utils.test.js [hunk 8]
```diff
@@ -163,6 +183,7 @@ describe('Utils functions', () => {
       expect(mockElements[1].focus.callCount).to.equal(0);
       expect(mockElements[2].focus.callCount).to.equal(1);
       expect(mockElements[3].focus.callCount).to.equal(0);
+
```
Comment (ktp242) [inline test/unit/utils.test.js]: I don't think we need an additional line here?

**Comment by ktp242** [comment]
Small syntax stuff! But it looks good!
